### ISSUE TITLE
Warn about unquoted use of identifiers bound under TYPE_CHECKING

### ISSUE THREAD

**Issue by sin-ack** [microsoft/pyright#5756]
**Is your feature request related to a problem? Please describe.**

This typechecks under Pyright 1.1.323, but fails at runtime:
```py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from decimal import Decimal

class Foo:
    bar: Decimal
```
The problem is that `Decimal` doesn't actually exist at runtime because `TYPE_CHECKING` is defined to be always `False`. This can get especially annoying when the type annotation is under a function that's executed way down the line, since the import and the code using it can be very far apart, leading to having to jump back and forth between the top of the file and the code to make sure you're not accidentally forgetting to quote.

**Describe the solution you’d like**
It would be great if Pyright could warn you about using quotes here (or even better, offer to insert them for you with a code action/during autocomplete). Without looking into Pyright internals, the implementation seems rather simple:

1. Collect all bindings within the module with `TYPE_CHECKING` assumed to be `False`.
2. Do the same but with `TYPE_CHECKING` assumed to be `True` (could be part of the regular type check).
3. If an identifier that's in (2) but not (1) is used unquoted, raise a warning about it.

This should cover the most common mistakes and save some headaches.

**Comment by erictraut** [comment]
The `TYPE_CHECKING` symbol is designed as a way to "lie" to a type checker about what is being executed at runtime. This makes it suitable for working around certain situations that cannot otherwise be expressed at runtime, but it also makes it very dangerous to use in your code. It's much safer if the type checker analyzes the same code you intend to run. For that reason, I recommend against using `TYPE_CHECKING` unless there's no alternative.

Making the type checker's behavior change based on whether statements are within a `TYPE_CHECKING` block goes against the very purpose of `TYPE_CHECKING`. So what you're suggesting in this enhancement request goes against a principle that I don't intend to break. Pyright's current way of handling `TYPE_CHECKING` is by design (and in accordance with PEP 484). Changing the behavior would not only go against the purpose of `TYPE_CHECKING`, but it would also potentially encourage the use of `TYPE_CHECKING` in cases for which it is not appropriate and even dangerous.

The solution to the problem you mention above is to eliminate your use of `TYPE_CHECKING`. In your example above, you have placed an `import` statement within a `TYPE_CHECKING` block, so you're lying to the type checker and telling it that you want it to assume that's going to be there at runtime. As you note, this can quickly lead to problems.

There are two reasons I've heard users say that they want to place `include` statements under a `TYPE_CHECKING` conditional. The first is to eliminate the performance overhead of an import. If your code is making use of a class (like `Decimal` in your example), then the module that defines this class will necessarily already be loaded, so importing it again is very cheap. So this argument doesn't make sense.

The second reason is that they need to use `TYPE_CHECKING` to eliminate a circular import dependency. A circularity is a sign that your code is not layered or structured properly. It should be refactored so it is properly layered. There's almost always a straightforward refactoring that will eliminate the circularity entirely and make your code more maintainable in the process. So this isn't a good excuse for using `TYPE_CHECKING`.

**Comment by sin-ack** [comment]
A third use-case is avoiding importing classes which require other things to be initialized first. At work we use Django. Django models require `apps.populate()` to run before they can be imported. Therefore, if we want to use things like `AbstractBaseUser` in our types we must use `TYPE_CHECKING`. This can frequently lead to problems in code where you want to do stuff like `cast(AbstractBaseUser, request.user)` (i.e. in code which expects authenticated requests only) because you now can be hit with a runtime error only when the code actually runs.

Another (admittedly minor) one is with things like `django-stubs` which expose types that only exist in Django as runtime-generated classes (like `django.db.models.RelatedManager`, which doesn't exist as a concrete class but is generated for one-to-many reverse relationships on model classes).

FWIW, I'm not asking for the behavior of the type-checking part of Pyright to be changed, but rather `pyright-language-server` (sorry if that wasn't clear). I hope you re-consider this request.

**Comment by erictraut** [comment]
The diagnostics reported by the language server are the same as the diagnostics reported by the command-line version of pyright. Breaking this rule would result in different diagnostics between the language server and CI scripts — something that we never want to do. So no, we wouldn't consider making this behavior specific to the language server.

As I said above, there may be rare cases where `TYPE_CHECKING` is required, but in my experience, the vast majority of its uses are unnecessary. It is often employed out of expediency by developers who don't understand the dangerous implications of its use.

Better documentation could perhaps help inform developers of these dangers and offer alternative solutions that are more sound.

Could you provide a minimal code sample that demonstrates the problem with django? I'd like to understand whether `TYPE_CHECKING` really is needed in this case or whether I can come up with a better approach.

**Comment by sin-ack** [comment]
Sure, here's an example of the latter in my last message:

```py
# add to `myapp/models.py` of an app after `django-admin startproject myproject && cd myproject && django-admin startapp myapp`.
# Must have `django-stubs` installed. 
from django.db import models

class Foo(models.Model):
    name = models.CharField[str, str](max_length=20)
    
    # Runtime error here, but no error from Pyright, because `django-stubs` declares it:
    bars: models.RelatedManager["Bar"]


class Bar(models.Model):
    name = models.CharField[str, str](max_length=20)
    foo = models.ForeignKey["Foo", "Foo"](to="myapp.Foo", on_delete=models.CASCADE)
```

The former is much more situational I'll admit, but it matters in cases where the following events happen during Django initialization:
1. Some modules like Django REST Framework accept _class paths_ (`path.to.import.ClassName`) for certain settings in `settings.py`. These are imported with `importlib` when Django is starting up.
2. At that point the apps haven't finished populating yet, so models can't be imported.
3. The code that's in the setting needs to use the model during a request. It's technically valid code, but because the model is imported during apps population, it results in a runtime error due to the metaclass magic that Django does. Therefore in order to get types, you must import with `TYPE_CHECKING`.

**Comment by sin-ack** [comment]
We're using a fairly recent version in the project where we had the issue I think, but regardless, I'll try to create a repository displaying the problem for you in a day or two.

The metaclass magic is unlikely to go away, so I understand if it's out of scope. It's just kind of unfortunate.
